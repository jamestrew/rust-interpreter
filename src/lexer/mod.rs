mod tokens;

use std::str::{from_utf8, Utf8Error};

use thiserror::Error;
pub use tokens::{Token, TokenKind};

#[derive(Error, Debug)]
enum TokenizerError {
    #[error("Unexpected input {0}. Failed to tokenize.")]
    UnexpectedInput(char),

    #[error("Invalid non UTF-8 string found: {0}")]
    NonUTF8Input(Utf8Error),
}

type Result<T> = std::result::Result<T, TokenizerError>;

#[derive(Debug, Default)]
pub struct Lexer {
    input: Vec<u8>,
    position: usize,
    read_position: usize,
    ch: u8,
    line: usize,
    col: usize,
}

impl Lexer {
    pub fn new(input: &str) -> Self {
        let mut lexer = Self {
            input: input.as_bytes().to_vec(),
            ..Default::default()
        };
        lexer.read_char();
        lexer.col = 0;
        lexer
    }

    fn read_char(&mut self) {
        if self.read_position >= self.input.len() {
            self.ch = 0;
        } else {
            self.ch = self.input[self.read_position];
        }
        self.position = self.read_position;
        self.read_position += 1;
        self.col += 1;
    }

    fn next_token(&mut self) -> Result<Token> {
        self.skip_whitespace();

        if self.is_letter() {
            return self.get_ident_or_kw();
        }

        let span_start = self.col;
        let token_type = match self.ch {
            b'=' => self.if_peek(b'=', TokenKind::Equal, TokenKind::Assign),
            b'+' => TokenKind::Plus,
            b'-' => TokenKind::Minus,
            b'*' => TokenKind::Asterisk,
            b'/' => TokenKind::ForwardSlash,
            b'!' => self.if_peek(b'=', TokenKind::NotEqual, TokenKind::Bang),

            b'<' => TokenKind::LT,
            b'>' => TokenKind::GT,

            b',' => TokenKind::Comma,
            b';' => TokenKind::Semicolon,
            b':' => TokenKind::Colon,

            b'(' => TokenKind::LParen,
            b')' => TokenKind::RParen,
            b'{' => TokenKind::LBrace,
            b'}' => TokenKind::RBrace,
            b'[' => TokenKind::LBracket,
            b']' => TokenKind::RBracket,

            b'0'..=b'9' => return self.get_int(),
            b'"' => return self.get_str(),

            0 => TokenKind::Eof,
            _ => return Err(TokenizerError::UnexpectedInput(self.ch as char)),
        };

        self.read_char();
        let span_end = self.col;
        Ok(Token::new(token_type, self.line, span_start..span_end))
    }

    fn skip_whitespace(&mut self) {
        while self.ch.is_ascii_whitespace() {
            let last_char = self.ch;
            self.read_char();
            if last_char == b'\n' {
                self.line += 1;
                self.col = 0;
            }
        }
    }

    fn is_letter(&self) -> bool {
        matches!(self.ch, b'a'..=b'z' | b'A'..=b'Z' | b'_' )
    }

    fn get_ident_or_kw(&mut self) -> Result<Token> {
        let span_start = self.col;
        let ident = self.read_ident()?;
        let token_type = match ident {
            "let" => TokenKind::Let,
            "fn" => TokenKind::Function,
            "if" => TokenKind::If,
            "else" => TokenKind::Else,
            "true" => TokenKind::True,
            "false" => TokenKind::False,
            "return" => TokenKind::Return,
            "nil" => TokenKind::Nil,
            _ => TokenKind::Identifier(ident.into()),
        };

        let span_end = self.col;

        Ok(Token::new(token_type, self.line, span_start..span_end))
    }

    fn read_ident(&mut self) -> Result<&str> {
        let start = self.position;
        while self.is_letter() {
            self.read_char()
        }
        from_utf8(&self.input[start..self.position]).map_err(TokenizerError::NonUTF8Input)
    }

    fn get_int(&mut self) -> Result<Token> {
        let start = self.position;
        let span_start = self.col;
        while self.ch.is_ascii_digit() {
            self.read_char();
        }

        let span_end = self.col;
        let val =
            from_utf8(&self.input[start..self.position]).map_err(TokenizerError::NonUTF8Input)?;
        Ok(Token::new(
            TokenKind::Int(val.into()),
            self.line,
            span_start..span_end,
        ))
    }

    fn get_str(&mut self) -> Result<Token> {
        let span_start = self.col;
        self.read_char();
        let start = self.position;
        while self.ch != b'"' {
            self.read_char();
        }
        let val =
            from_utf8(&self.input[start..self.position]).map_err(TokenizerError::NonUTF8Input)?;
        let token_type = TokenKind::Str(val.into());
        self.read_char();
        let span_end = self.col;
        Ok(Token::new(token_type, self.line, span_start..span_end))
    }

    fn if_peek(&mut self, peek: u8, true_token: TokenKind, false_token: TokenKind) -> TokenKind {
        if let Some(token) = self.peek() {
            if token == peek {
                self.read_char();
                return true_token;
            }
        }
        false_token
    }

    fn peek(&self) -> Option<u8> {
        self.input.get(self.read_position).copied()
    }
}

impl Iterator for Lexer {
    type Item = Token;

    fn next(&mut self) -> Option<Self::Item> {
        let token = self.next_token().ok()?;

        if *token.kind() == TokenKind::Eof && self.position - 1 > self.input.len() {
            None
        } else {
            Some(token)
        }
    }
}

#[cfg(test)]
mod test {
    use super::*;

    #[allow(unused)]
    #[derive(Debug)]
    struct T<'a> {
        token: Token,
        slice: &'a str,
    }

    fn lex(input: &str) -> Vec<Token> {
        let lexer = Lexer::new(input);

        lexer.collect()
    }

    fn debug_print(input: &str, tokens: Vec<Token>) -> Vec<T> {
        let lines = input.lines().collect::<Vec<_>>();

        let mut ret = Vec::new();
        for token in tokens {
            let line = lines[token.line];
            let slice = &line[token.span.clone()];
            ret.push(T { token, slice });
        }
        ret
    }

    macro_rules! snapshot {
        ($name:tt, $input:expr) => {
            #[test]
            fn $name() {
                let tokens = lex($input);
                insta::with_settings!({
                    description => $input,
                }, {
                    insta::assert_debug_snapshot!(debug_print($input, tokens));
                })
            }
        };
    }

    snapshot!(basic, "=");
    snapshot!(symbols, "=+(){},;");
    snapshot!(assignments1, "let five = 5;");
    snapshot!(assignments2, "let ten = 10;");
    snapshot!(
        assignments3,
        r"let add = fn(x, y) {
        x + y;
    };"
    );
    snapshot!(assignments4, "let result = add(five, ten);");

    snapshot!(operators, "!-/*5;");
    snapshot!(equality1, "5 < 10 > 5;");
    snapshot!(equality2, "10 == 10;");
    snapshot!(equality3, "10 != 9;");
    snapshot!(
        conditional,
        r"if (5 < 10) {
        return true;
    } else {
        return false;
    }"
    );

    snapshot!(string_literal, r#""hello world";"#);
    snapshot!(arrays, "[1, 2]");
    snapshot!(dictionary, r#"{ "foo": "bar" };"#);
}
