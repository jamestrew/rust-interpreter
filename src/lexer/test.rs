use super::*;
use crate::errors::MonkeyError;

struct T<'a> {
    token: TokenKind,
    slice: &'a str,
}

impl<'a> std::fmt::Debug for T<'a> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "(token: '{}', slice: '{}')", self.token, self.slice)
    }
}

enum K<'a> {
    Token(T<'a>),
    Error(MonkeyError),
}

impl<'a> std::fmt::Debug for K<'a> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Token(t) => write!(f, "Token{:?}", t),
            Self::Error(err) => write!(f, "Error({})", err),
        }
    }
}

fn lex(input: &str) -> Vec<Result<Token>> {
    let lexer = Lexer::new(input);
    lexer.collect()
}

fn debug_print(input: &str, token_results: Vec<Result<Token>>) -> Vec<K> {
    let rows = input.lines().collect::<Vec<_>>();

    let mut ret = Vec::new();
    for result in token_results {
        match result {
            Ok(token) => {
                println!("{:?}", token);
                let row_num = rows[token.row_num()];
                let slice = &row_num[token.span()];
                ret.push(K::Token(T {
                    token: token.take(),
                    slice,
                }));
            }
            Err(err) => ret.push(K::Error(err.take())),
        }
    }
    ret
}

macro_rules! snapshot {
    ($name:tt, $input:expr) => {
        #[test]
        fn $name() {
            let tokens = lex($input);
            insta::with_settings!({
                description => $input,
            }, {
                insta::assert_debug_snapshot!(debug_print($input, tokens));
            })
        }
    };
}

snapshot!(basic, "=");
snapshot!(symbols, "=+(){},;");
snapshot!(assignments1, "let five = 5;");
snapshot!(assignments2, "let ten = 10;");
snapshot!(
    assignments3,
    r"let add = fn(x, y) {
        x + y;
    };"
);
snapshot!(assignments4, "let result = add(five, ten);");

snapshot!(operators, "!-/*5;");
snapshot!(equality1, "5 < 10 > 5;");
snapshot!(equality2, "10 == 10;");
snapshot!(equality3, "10 != 9;");
snapshot!(
    conditional,
    r"if (5 < 10) {
        return true;
    } else {
        return false;
    }"
);

snapshot!(string_literal, r#""hello world";"#);
snapshot!(arrays, "[1, 2]");
snapshot!(dictionary, r#"{ "foo": "bar" };"#);
