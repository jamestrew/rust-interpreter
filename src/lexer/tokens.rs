use std::rc::Rc;

#[derive(Debug, Clone, PartialEq)]
pub enum Token {
    Assign,
    Plus,
    Minus,
    Asterisk,
    ForwardSlash,
    Bang,

    Equal,
    NotEqual,
    LT,
    GT,

    Comma,
    Semicolon,
    Colon,

    LParen,
    RParen,
    LBrace,
    RBrace,
    LBracket,
    RBracket,

    Identifier(Rc<str>),
    Int(Rc<str>),
    Str(Rc<str>),

    // keywords
    Let,
    Function,
    If,
    Else,
    True,
    False,
    Return,

    Eof,
}

impl AsRef<Token> for Token {
    fn as_ref(&self) -> &Token {
        self
    }
}

impl std::fmt::Display for Token {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Token::Assign => write!(f, "="),
            Token::Plus => write!(f, "+"),
            Token::Minus => write!(f, "-"),
            Token::Asterisk => write!(f, "*"),
            Token::ForwardSlash => write!(f, "/"),
            Token::Bang => write!(f, "!"),
            Token::Equal => write!(f, "=="),
            Token::NotEqual => write!(f, "!="),
            Token::LT => write!(f, "<"),
            Token::GT => write!(f, ">"),
            Token::Comma => write!(f, ","),
            Token::Semicolon => write!(f, ";"),
            Token::Colon => write!(f, ":"),
            Token::LParen => write!(f, "("),
            Token::RParen => write!(f, ")"),
            Token::LBrace => write!(f, "{{"),
            Token::RBrace => write!(f, "}}"),
            Token::LBracket => write!(f, "["),
            Token::RBracket => write!(f, "]"),
            Token::Identifier(ident) => write!(f, "{}", ident),
            Token::Int(val) => write!(f, "{}", val),
            Token::Str(val) => write!(f, "\"{}\"", val),
            Token::Let => write!(f, "let"),
            Token::Function => write!(f, "fn"),
            Token::If => write!(f, "if"),
            Token::Else => write!(f, "else"),
            Token::True => write!(f, "true"),
            Token::False => write!(f, "false"),
            Token::Return => write!(f, "return"),
            Token::Eof => write!(f, "EOF"),
        }
    }
}
