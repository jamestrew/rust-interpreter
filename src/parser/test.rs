use crate::parse;

macro_rules! assert_stmt {
    ($name:tt, $input:expr) => {
        #[test]
        fn $name() {
            let stmts = parse($input);
            let stmt = stmts[0].to_string();
            assert_eq!($input, stmt);
        }
    };
}

macro_rules! ss_display_stmt {
    ($name:tt, $input:expr) => {
        #[test]
        fn $name() {
            let statements = parse($input);
            let first = &statements[0];
            insta::with_settings!({
                description => $input,
            }, {
                insta::assert_display_snapshot!(first);
            })
        }
    };
}

macro_rules! ss_debug_stmts {
    ($name:tt, $input:expr) => {
        #[test]
        fn $name() {
            let stmts = parse($input);
            insta::with_settings!({
                description => $input,
            }, {
                insta::assert_debug_snapshot!(stmts);
            })
        }
    };
}

#[test]
fn swallow_extra_semicolons() {
    let stmts = parse("return true;;;;");
    let stmt = stmts[0].to_string();
    assert_eq!(stmt, "return true;");

    let statements = parse("foobar;;;;");
    let first = statements[0].to_string();
    assert_eq!(first, "foobar;");
}

assert_stmt!(literal_let_statement_1, "let x = 5;");
assert_stmt!(literal_let_statement_2, "let y = 10;");
assert_stmt!(literal_let_statement_3, "let foobar = 838383;");
assert_stmt!(literal_let_statement_4, "let foo = \"bar\";");
assert_stmt!(literal_let_statement_5, "let foo = true;");
assert_stmt!(identifier_let_statement, "let foo = foobar;");

assert_stmt!(literal_return_statement_1, "return 5;");
assert_stmt!(literal_return_statement_2, "return true;");
assert_stmt!(literal_return_statement_3, "return \"foo\";");
assert_stmt!(identifier_return_statement, "return foobar;");

assert_stmt!(identifier, "foobar;");
assert_stmt!(integer_literal, "1;");
assert_stmt!(boolean_literal, "true;");
assert_stmt!(string_literal, "\"hello world\";");

ss_display_stmt!(prefix_expression_1, "-5;");
ss_display_stmt!(prefix_expression_2, "!foobar;");

ss_display_stmt!(infix_expr_1, "5 + 6;");
ss_display_stmt!(infix_expr_2, "5 - 6;");
ss_display_stmt!(infix_expr_3, "5 * 6;");
ss_display_stmt!(infix_expr_4, "5 / 6;");
ss_display_stmt!(infix_expr_5, "true == true;");
ss_display_stmt!(infix_expr_6, "true != false;");
ss_display_stmt!(infix_expr_7, "5 < 6");
ss_display_stmt!(infix_expr_8, "7 > 6");
ss_display_stmt!(infix_expr_9, "\"foo\" != \"bar\"");

ss_display_stmt!(operator_precedence_1, "-a * b");
ss_display_stmt!(operator_precedence_2, "!-a");
ss_display_stmt!(operator_precedence_3, "a + b + c");
ss_display_stmt!(operator_precedence_4, "a + b - c");
ss_display_stmt!(operator_precedence_5, "a * b * c");
ss_display_stmt!(operator_precedence_6, "a * b / c");
ss_display_stmt!(operator_precedence_7, "a + b / c");
ss_display_stmt!(operator_precedence_8, "a + b * c + d / e - f");
ss_display_stmt!(operator_precedence_9, "5 > 4 == 3 < 4");
ss_display_stmt!(operator_precedence_10, "5 < 4 != 3 > 4");
ss_display_stmt!(operator_precedence_11, "3 + 4 * 5 == 3 * 1 + 4 * 5");
ss_display_stmt!(operator_precedence_12, "true");
ss_display_stmt!(operator_precedence_13, "false");
ss_display_stmt!(operator_precedence_14, "3 > 5 == false");
ss_display_stmt!(operator_precedence_15, "3 < 5 == true");
ss_display_stmt!(operator_precedence_16, "1 + (2 + 3) + 4");
ss_display_stmt!(operator_precedence_17, "(5 + 5) * 2");
ss_display_stmt!(operator_precedence_18, "2 / (5 + 5)");
ss_display_stmt!(operator_precedence_19, "-(5 + 5)");
ss_display_stmt!(operator_precedence_20, "!(true == true)");
// ss_display_stmt!(operator_precedence_21, "a + add(b * c) + d");
// ss_display_stmt!(
//     operator_precedence_22,
//     "add(a, b, 1, 2 * 3, 4 + 5, add(6, 7 * 8))"
// );
// ss_display_stmt!(operator_precedence_23, "add(a + b + c * d / f + g)");
// ss_display_stmt!(operator_precedence_24, "a * [1, 2, 3, 4][b * c] * d");
// ss_display_stmt!(operator_precedence_25, "add(a * b[2], b[1], 2 * [1, 2][1])");

ss_display_stmt!(if_expr_1, "if (x < y) { x }");
ss_display_stmt!(if_expr_2, "if (x < y) { x } else { y }");
ss_display_stmt!(if_expr_3, "if (x < y) { x } else { let z = x + y; z }");

ss_debug_stmts!(block_1, "{ return 5; }");
ss_debug_stmts!(block_2, "{ return 5; return true; }");
