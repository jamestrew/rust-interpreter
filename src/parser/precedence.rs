#![allow(dead_code)]
use crate::lexer::Token;

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
pub enum Precedence {
    Lowest,
    Equals,
    LessGreater,
    Sum,
    Product,
    Prefix,
    Call,
    Index,
}

impl From<&Token> for Precedence {
    fn from(token: &Token) -> Self {
        match token {
            Token::Equal | Token::NotEqual => Self::Equals,
            Token::LT | Token::GT => Self::LessGreater,
            Token::Plus | Token::Minus => Self::Sum,
            Token::Asterisk | Token::ForwardSlash => Self::Product,
            Token::LParen => Self::Call,
            Token::LBracket => Self::Index,
            _ => Self::Lowest,
        }
    }
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn precedence_order() {
        assert!(Precedence::Lowest < Precedence::Equals);
        assert!(Precedence::Equals < Precedence::LessGreater);
        assert!(Precedence::LessGreater < Precedence::Sum);
        assert!(Precedence::Sum < Precedence::Product);
        assert!(Precedence::Product < Precedence::Prefix);
        assert!(Precedence::Prefix < Precedence::Call);
        assert!(Precedence::Call < Precedence::Index);
    }
}
