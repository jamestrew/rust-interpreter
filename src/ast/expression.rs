use std::fmt::{Debug, Display};
use std::rc::Rc;

use super::Node;
use crate::lexer::Token;
use crate::parser::*;

#[derive(PartialEq, Clone)]
pub enum Expression {
    Identifier(Identifier),
    Primative(Primative),
    StringLiteral(Rc<str>),
    Prefix(Prefix),
    Infix(Infix),
}

impl Node for Expression {}

impl Expression {
    pub fn parse(parser: &mut Parser, precedence: Precedence) -> anyhow::Result<Self> {
        let token = parser.current_token()?;
        let mut expr = match token {
            Token::Int(_) | Token::True | Token::False => {
                Expression::Primative(Primative::parse(parser)?)
            }
            Token::Str(s) => Expression::StringLiteral(s.clone()),
            Token::Identifier(_) => Expression::Identifier(Identifier::parse(parser)?),
            Token::Minus | Token::Bang => Expression::Prefix(Prefix::parse(parser)?),
            _ => todo!("Expression::parse for {:?}", token),
        };

        while precedence < parser.peek_precedence()? {
            expr = Expression::Infix(Infix::parse(parser, expr)?);
        }
        parser.swallow_semicolons();
        Ok(expr)
    }
}

impl Debug for Expression {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Expression::Identifier(val) => write!(f, "{:?}", val),
            Expression::Primative(val) => write!(f, "{:?}", val),
            Expression::StringLiteral(val) => write!(f, "StringLiteral({:?})", val),
            Expression::Prefix(val) => write!(f, "{:?}", val),
            Expression::Infix(val) => write!(f, "{:?}", val),
        }
    }
}

impl Display for Expression {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let s = match self {
            Expression::Identifier(val) => val.to_string(),
            Expression::Primative(val) => val.to_string(),
            Expression::StringLiteral(val) => format!("\"{}\"", val),
            Expression::Prefix(val) => val.to_string(),
            Expression::Infix(val) => val.to_string(),
        };
        write!(f, "{}", s)
    }
}

#[derive(PartialEq, Clone)]
pub struct Identifier {
    token: Token,
    value: Rc<str>,
}

impl Node for Identifier {}

impl Identifier {
    pub fn parse(parser: &Parser) -> anyhow::Result<Self> {
        Self::try_from(parser.current_token()?)
    }
}

impl<'a> TryFrom<&'a Token> for Identifier {
    type Error = anyhow::Error;

    fn try_from(token: &'a Token) -> anyhow::Result<Self> {
        if let Token::Identifier(ident) = token {
            Ok(Self {
                token: token.clone(),
                value: ident.clone(),
            })
        } else {
            Err(anyhow::anyhow!(
                "Identifier can only be created from an Identifier token"
            ))
        }
    }
}

impl Debug for Identifier {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "Identifier({:?})", self.value)
    }
}

impl Display for Identifier {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.value)
    }
}

#[derive(PartialEq, Clone)]
pub enum Primative {
    Int(i64),
    Bool(bool),
}

impl Node for Primative {}

impl Primative {
    pub fn parse(parser: &Parser) -> anyhow::Result<Self> {
        let token = parser.current_token()?;
        match token {
            Token::Int(val) => Ok(Self::Int(val.parse::<i64>()?)),
            Token::True => Ok(Self::Bool(true)),
            Token::False => Ok(Self::Bool(false)),
            _ => unreachable!("Primative parse unexpected {:?}", token),
        }
    }
}

impl Debug for Primative {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Int(val) => write!(f, "Int({val})"),
            Self::Bool(val) => write!(f, "Bool({val})"),
        }
    }
}

impl Display for Primative {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Primative::Int(val) => write!(f, "{}", val),
            Primative::Bool(val) => write!(f, "{}", val),
        }
    }
}

#[derive(PartialEq, Clone)]
pub struct Prefix {
    token: Token,
    right: Box<Expression>,
}

impl Node for Prefix {}

impl Prefix {
    pub fn parse(parser: &mut Parser) -> anyhow::Result<Self> {
        let operator = parser.current_token()?.clone();
        parser.next_token();
        let right = Expression::parse(parser, Precedence::Prefix)?;
        Ok(Self {
            token: operator,
            right: Box::new(right),
        })
    }
    fn operator_str(&self) -> &'static str {
        match self.token {
            Token::Minus => "-",
            Token::Bang => "!",
            _ => unreachable!("only Bang and Minus are allowed prefixes"),
        }
    }
}

impl Debug for Prefix {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("Prefix")
            .field("token", &self.operator_str())
            .field("right", &self.right)
            .finish()
    }
}

impl Display for Prefix {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "({}{})", self.operator_str(), self.right)
    }
}

#[derive(PartialEq, Clone)]
pub struct Infix {
    token: Token,
    left: Box<Expression>,
    right: Box<Expression>,
}

impl Node for Infix {}

impl Infix {
    fn operator_str(&self) -> &'static str {
        match self.token {
            Token::Minus => "-",
            Token::Plus => "+",
            Token::Asterisk => "*",
            Token::ForwardSlash => "/",
            Token::Equal => "==",
            Token::NotEqual => "!=",
            Token::LT => "<",
            Token::GT => ">",
            _ => unreachable!("unallowed infix operator {:?}", self.token),
        }
    }

    fn parse(parser: &mut Parser, left: Expression) -> anyhow::Result<Self> {
        parser.next_token();
        let op_token = parser.current_token()?.clone();
        let op_precedence = parser.current_precedence()?;
        parser.next_token();
        let right = Expression::parse(parser, op_precedence)?;
        Ok(Self {
            token: op_token,
            left: Box::new(left),
            right: Box::new(right),
        })
    }
}

impl Debug for Infix {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("Infix")
            .field("token", &self.operator_str())
            .field("left", &self.left)
            .field("right", &self.right)
            .finish()
    }
}

impl Display for Infix {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "({} {} {})", self.left, self.operator_str(), self.right)
    }
}

#[cfg(test)]
mod test {
    use crate::ast::Statement;
    use crate::lexer::Lexer;
    use crate::parser::Parser;

    fn parse(input: &str) -> Vec<Statement> {
        let lexer = Lexer::new(input);
        let mut parser = Parser::new(lexer);
        let program = parser.parse_programe().expect("valid program");
        assert!(!program.statements.is_empty());
        program.statements
    }

    #[test]
    fn swallow_extra_semicolons() {
        let statements = parse("foobar;;;;");
        let first = statements[0].to_string();
        assert_eq!(first, "foobar;");
    }

    macro_rules! test {
        ($name:tt, $input:expr) => {
            #[test]
            fn $name() {
                let statements = parse($input);
                let first = statements[0].to_string();
                assert_eq!($input, first);
            }
        };
    }

    macro_rules! snapshot {
        ($name:tt, $input:expr) => {
            #[test]
            fn $name() {
                let statements = parse($input);
                let first = &statements[0];
                insta::with_settings!({
                    description => $input,
                }, {
                    insta::assert_display_snapshot!(first);
                })
            }
        };
    }

    macro_rules! snapshot_debug {
        ($name:tt, $input:expr) => {
            #[test]
            fn $name() {
                let statements = parse($input);
                let first = &statements[0];
                insta::with_settings!({
                    description => $input,
                }, {
                    insta::assert_debug_snapshot!(first);
                })
            }
        };
    }

    test!(identifier, "foobar;");
    test!(integer_literal, "1;");
    test!(boolean_literal, "true;");
    test!(string_literal, "\"hello world\";");

    snapshot_debug!(prefix_expression_1, "-5;");
    snapshot_debug!(prefix_expression_2, "!foobar;");

    snapshot_debug!(infix_expr_1, "5 + 6;");
    snapshot_debug!(infix_expr_2, "5 - 6;");
    snapshot_debug!(infix_expr_3, "5 * 6;");
    snapshot_debug!(infix_expr_4, "5 / 6;");
    snapshot_debug!(infix_expr_5, "true == true;");
    snapshot_debug!(infix_expr_6, "true != false;");
    snapshot_debug!(infix_expr_7, "5 < 6");
    snapshot_debug!(infix_expr_8, "7 > 6");
    snapshot_debug!(infix_expr_9, "\"foo\" != \"bar\"");

    snapshot!(operator_precedence_1, "-a * b");
    snapshot!(operator_precedence_2, "!-a");
    snapshot!(operator_precedence_3, "a + b + c");
    snapshot!(operator_precedence_4, "a + b - c");
    snapshot!(operator_precedence_5, "a * b * c");
    snapshot!(operator_precedence_6, "a * b / c");
    snapshot!(operator_precedence_7, "a + b / c");
    snapshot!(operator_precedence_8, "a + b * c + d / e - f");
    snapshot!(operator_precedence_9, "5 > 4 == 3 < 4");
    snapshot!(operator_precedence_10, "5 < 4 != 3 > 4");
    snapshot!(operator_precedence_11, "3 + 4 * 5 == 3 * 1 + 4 * 5");
    snapshot!(operator_precedence_12, "true");
    snapshot!(operator_precedence_13, "false");
    snapshot!(operator_precedence_14, "3 > 5 == false");
    snapshot!(operator_precedence_15, "3 < 5 == true");
    // snapshot!(operator_precedence_16, "1 + (2 + 3) + 4");
    // snapshot!(operator_precedence_17, "(5 + 5) * 2");
    // snapshot!(operator_precedence_18, "2 / (5 + 5)");
    // snapshot!(operator_precedence_19, "-(5 + 5)");
    // snapshot!(operator_precedence_20, "!(true == true)");
    // snapshot!(operator_precedence_21, "a + add(b * c) + d");
    // snapshot!(
    //     operator_precedence_22,
    //     "add(a, b, 1, 2 * 3, 4 + 5, add(6, 7 * 8))"
    // );
    // snapshot!(operator_precedence_23, "add(a + b + c * d / f + g)");
    // snapshot!(operator_precedence_24, "a * [1, 2, 3, 4][b * c] * d");
    // snapshot!(operator_precedence_25, "add(a * b[2], b[1], 2 * [1, 2][1])");
}
