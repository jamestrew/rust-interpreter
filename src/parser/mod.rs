mod precedence;

use crate::ast::{Node, Program, Statement};
use crate::lexer::{Lexer, Token};
pub use precedence::Precedence;

pub struct Parser {
    lexer: Lexer,
    current_token: Option<Token>,
    peek_token: Option<Token>,
}

impl Parser {
    pub fn new(mut lexer: Lexer) -> Self {
        let current_token = lexer.next();
        let peek_token = lexer.next();
        Self {
            lexer,
            current_token,
            peek_token,
        }
    }

    pub fn parse_programe(&mut self) -> anyhow::Result<Program> {
        let mut program = Program::default();
        while !self.current_token_is(Token::Eof) {
            program.statements.push(Statement::parse(self, None)?);
            self.next_token();
        }
        Ok(program)
    }

    pub fn current_token_is(&self, match_token: Token) -> bool {
        if let Some(token) = &self.current_token {
            *token == match_token
        } else {
            false
        }
    }

    pub fn peek_token_is(&self, match_token: Token) -> bool {
        if let Some(token) = &self.peek_token {
            *token == match_token
        } else {
            false
        }
    }

    pub fn current_token(&self) -> anyhow::Result<&Token> {
        if self.current_token.is_none() {
            return Err(anyhow::anyhow!("No current token"));
        }
        Ok(self.current_token.as_ref().unwrap())
    }

    pub fn peek_token(&self) -> anyhow::Result<&Token> {
        if self.peek_token.is_none() {
            return Err(anyhow::anyhow!("No peek token"));
        }
        Ok(self.peek_token.as_ref().unwrap())
    }

    pub fn next_token(&mut self) {
        self.current_token = self.peek_token.take();
        self.peek_token = self.lexer.next();
    }

    pub fn swallow_semicolons(&mut self) {
        while self.peek_token_is(Token::Semicolon) {
            self.next_token();
        }
    }
}
