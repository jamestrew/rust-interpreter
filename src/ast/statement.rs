use anyhow::anyhow;

use super::expression::{Expression, Identifier};
use super::Node;
use crate::lexer::Token;
use crate::parser::Parser;

#[derive(Debug, PartialEq)]
pub enum Statement {
    Let(Let),
    // ...
}

impl Node for Statement {
    fn parse(_parser: &mut Parser) -> anyhow::Result<Self>
    where
        Self: std::marker::Sized,
    {
        todo!()
    }
}

impl std::fmt::Display for Statement {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let s = match self {
            Statement::Let(val) => val.to_string(),
        };
        write!(f, "{}", s)
    }
}

#[derive(Debug, PartialEq)]
pub struct Let {
    token: Token,
    name: Identifier,
    value: Expression,
}

impl Let {
    pub fn new(name: Identifier, value: Expression) -> Self {
        Self {
            token: Token::Let,
            name,
            value,
        }
    }
}

impl Node for Let {
    fn parse(parser: &mut Parser) -> anyhow::Result<Self>
    where
        Self: std::marker::Sized,
    {
        parser.next_token();
        let name = Identifier::parse(parser)?;
        parser.next_token();
        if parser.current_token_is(Token::Equal) {
            return Err(anyhow!("Expected `=` symbol in let statement"));
        }
        parser.next_token();
        let value = Expression::parse(parser)?;
        parser.next_token();
        Ok(Self {
            token: Token::Let,
            name,
            value,
        })
    }
}

impl std::fmt::Display for Let {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "let {} = {};", self.name, self.value)
    }
}

#[cfg(test)]
mod test {
    use crate::lexer::Lexer;
    use crate::parser::Parser;

    fn parse(input: &str) -> String {
        let lexer = Lexer::new(input);
        let mut parser = Parser::new(lexer);
        let program = parser.parse_programe().expect("valid program");
        assert!(!program.statements.is_empty());
        program.statements[0].to_string()
    }

    macro_rules! snapshot_first {
        ($name:tt, $input:expr) => {
            #[test]
            fn $name() {
                let output = parse($input);
                insta::assert_snapshot!(output, $input);
            }
        };
    }

    snapshot_first!(let_statement_1, "let x = 5;");
    snapshot_first!(let_statement_2, "let y = 10;");
    snapshot_first!(let_statement_3, "let foobar = 838383;");
    snapshot_first!(let_statement_4, "let foo = \"bar\";");
    snapshot_first!(let_statement_5, "let foo = true;");
}
