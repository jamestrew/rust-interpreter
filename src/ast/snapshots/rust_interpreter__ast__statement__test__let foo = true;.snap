---
source: src/ast/statement.rs
expression: "\"let foo = true;\""
---
let foo = true;
